### ISSUE TITLE
Implement missing torch.nan* operators

### ISSUE THREAD

**Issue by heitorschueroff** [pytorch/pytorch#61474]
As part of NumPy compatibility, we want to implement all remaining nan* operators such as torch.nanmean requested here https://github.com/pytorch/pytorch/issues/21987. 

It's been suggested [here](https://github.com/pytorch/pytorch/issues/46544#issuecomment-757556866) and [here](https://github.com/pytorch/pytorch/issues/21987#issuecomment-503756201) that we add a keyword argument for controlling the nan behavior, in which case the nan* operators would call the corresponding operator with the correct nan flag. See SciPy's design for [nan_policy](https://docs.scipy.org/doc/scipy/reference/dev/api-dev/nan_policy.html)

see also: https://github.com/pytorch/pytorch/issues/21987 https://github.com/pytorch/pytorch/issues/46544

## nan* operators to implement

- [ ] torch.nanmean
- [ ] torch.nanmax
- [ ] torch.nanmin
- [ ] torch.nanstd
- [ ] torch.nanvar
- [ ] torch.nanargmin
- [ ] torch.nanargmax
- [ ] torch.nanprod
- [ ] torch.nancumsum
- [ ] torch.nancumprod

**Comment by mruberry** [comment]
I like the idea of adding a nan_policy kwarg and implementing these as "partials" that set the nan_policy.

**Comment by SamsTheGreatest** [comment]
is the idea dead?

**Comment by mruberry** [comment]
> is the idea dead?

I think this is still the plan, but I don't believe anyone is actively working on it

**Comment by oiao** [comment]
The functionality is quite helpful when dealing with sparse data and it would be great to see this implemented 👍

**Comment by Linux-cpp-lisp** [comment]
Seconding interest in having this available, regardless of whether it's through `nan_policy` or `torch.nan*`; it's very helpful for masking data.

**Comment by PatrikBrynolfsson** [comment]
This functionality would be very useful! What is the status?

**Comment by kleinicke** [comment]
I frequently search for such a functionality, but Google only shows me torch.fmax as a solution, which is not helpful.
Sometimes I transformed the torch tensor to numpy to deal with this issue, but it's not really a good solution.

**Comment by MaxWolf-01** [comment]
nanstd would be very useful!

**Comment by PatrikBrynolfsson** [comment]
Hello! All nan-operators are very useful, and I miss them regularly. Please prioritize this!

**Comment by OlegPlatonov** [comment]
Hi! I also find nan* operators to be very useful and would like to see them in PyTorch.

Also, there are at least two very useful nan* operators that are available in NumPy but not mentioned in this issue: nanquantile and its special case nanmedian.

**Comment by mhollenberg** [comment]
I would also find more torch.nan functions very useful. I am normalizing my data prior to training and having a torch.nanstd would make things much easier!

**Comment by nbeuchat** [comment]
I would also love to see these operators implemented. If you have batches with `nan` padding, that's very handy to have these nan* operators, especially if the other ones are available.

**Comment by nbeuchat** [comment]
I'm using @amitabe's functions as a workaround but `nanvar` and `nanstd` do not work properly as they use the biased estimator while, by default, `std` uses the unbiased estimator (ie: the `correction` parameter)

Here is a version that takes it into account if that's important for you:

```
def nanvar(tensor, dim: int=None, keepdim: bool=False, correction: int=1):
    # Count non-NaN elements
    count = torch.sum(~torch.isnan(tensor), dim=dim, keepdim=keepdim)
    
    # Compute mean while ignoring NaNs
    mean = torch.nanmean(tensor, dim=dim, keepdim=True)
    
    # Compute squared difference from the mean
    sq_diff = (tensor - mean).pow(2)
    sq_diff = torch.where(torch.isnan(sq_diff), torch.tensor(0.0, device=tensor.device, dtype=tensor.dtype), sq_diff)
    
    # Sum of squared differences
    sum_sq_diff = sq_diff.sum(dim=dim, keepdim=keepdim)
    
    # Degrees of freedom correction (e.g., 1 for sample variance)
    divisor = count - correction
    divisor = torch.clamp(divisor, min=1)  # Avoid division by zero
    return sum_sq_diff / divisor

def nanstd(tensor, dim: int=None, keepdim: bool=False, correction: int=1):
    return nanvar(tensor, dim=dim, keepdim=keepdim, correction=correction).sqrt()
```
